dbt State doc updates (#9577)
fishtown-analytics/docs.getdbt.com: website/docs/faqs/State/views-rebuilt.md

Reviewer: It's avoiding an error - if you have this view 

```
select * from some_table
```

and some_table gains a column, then when the view is queried it will cause an error.

Author: @joellabes Clarified here:

```suggestion
dbt State reuses a model when its compiled SQL matches the stored hash. For views with `select *`, dbt State can't determine which columns the query selects without querying the upstream schema, so it can't confirm the SQL is unchanged. It always rebuilds these views to avoid errors &mdash; if the upstream table gains a column, querying the view can fail. When dbt State rebuilds a view, it also re-runs any tests defined on the model.
```

id: views-rebuilt
---

> dbt State decides whether to reuse a model by comparing its compiled SQL to the hash from the previous run. If anything changes that compiled SQL between runs, even when the logic hasn't meaningfully changed, dbt State rebuilds the model. In some cases, dbt State also rebuilds when it can't determine source freshness, regardless of SQL changes.

Reviewer 2: Can we add the why here:
Something like: The basis for state's decisions is that it is better to build (the previous behavior without state) than skip building when it's unclear if the data has changed.

Author: @reubenmc what do you think of:

```suggestion
dbt State decides whether to reuse a model by comparing its compiled SQL to the hash from the previous run. If anything changes that compiled SQL between runs, even when the logic hasn't meaningfully changed, dbt State rebuilds the model. In some cases, dbt State also rebuilds when it can't determine source freshness, regardless of SQL changes. This behavior is intentional: when it's unclear if the data has changed, dbt State prefers rebuilding the model (the same behavior as running without dbt State) over skipping a build that might be necessary.
```

Reviewer: >If anything changes that compiled SQL between runs, even when the logic hasn't meaningfully changed, dbt State rebuilds the model

This sounds counter to the idea of hashing the AST instead of the raw code. Any logic change will be picked up, but compiled SQL changes that don't impact logic will be ignored. (This handwaves over a lot, but is more accurate than not)

I'd say something like this: 

> dbt State decides whether to reuse a model by parsing the rendered SQL into a syntax tree and comparing the hash. If the hash has changed (implying the model's logic has changed), dbt State rebuilds the model.
> 
> dbt State prioritises safety and precision; if it can't guarantee skipping a node is safe, then it will re-execute to be sure. A few patterns that commonly cause overeager rebuilds are listed below, along with recommendations to increase reuse rate.

Author: Thanks @joellabes ! Applied some editorial changes here:

```suggestion
dbt State decides whether to reuse a model by parsing the rendered SQL into a syntax tree and comparing the hash. If the hash has changed (implying the model's logic has changed), dbt State rebuilds the model.

dbt State prioritizes safety and precision; if it can't guarantee skipping a node is safe, then it rebuilds the node to be sure. A few patterns that commonly cause overeager rebuilds are listed on this page, along with recommendations to increase reuse rate.
```
